### PR TITLE
Introduces http server crate

### PR DESCRIPTION
**Pull request by eadventurous**
Signed-off-by: Egor Ivkov <e.o.ivkov@gmail.com>

### CODE REVIEW

File: iroha_http_server/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,640 @@
...
+
+impl<State: Clone + Send + Sync + 'static> Server<State> {
+    pub fn new(state: State) -> Self {
+        Server {
+            state,
+            router: Arc::new(Router::new()),
+        }
+    }
+
+    pub fn at<'s>(&'s mut self, path: &str) -> RouteBuilder<'s, State> {
```
Comment (humb1t) [inline iroha_http_server/src/lib.rs:49]: Frustrating naming in my opinion, `route` may be better.

File: iroha_http_server/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,640 @@
...
+            let mut stream = stream.map_err(|e| e.to_string())?;
+            let router = self.router.clone();
+            let state = self.state.clone();
+            task::spawn(async move {
+                let mut buffer = [0u8; BUFFER_SIZE];
+                let read_size = stream
+                    .peek(&mut buffer)
+                    .await
+                    .expect("Request read failed.");
+                let response = match HttpRequest::try_from(&buffer[..]) {
```
Comment (humb1t) [inline iroha_http_server/src/lib.rs:70]: May be encapsulated in `let response = HttpRequest::try_from(...).process(...)`

File: iroha_http_server/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,640 @@
...
+        Ok(())
+    }
+}
+
+pub type WebSocketStream = TungsteniteWebSocketStream<TcpStream>;
+
+pub enum Endpoint<State> {
+    WebSocket(Box<dyn WebSocketHandler<State>>),
+    Get(Box<dyn HttpHandler<State>>),
+    Post(Box<dyn HttpHandler<State>>),
```
Comment (humb1t) [inline iroha_http_server/src/lib.rs:176]: for REST we will also need PUT, PATCH, DELETE, etc. - may be just depicted as TODO for now

Comment (humb1t) [reply iroha_http_server/src/lib.rs:176]: Or even better to make it another level
Endpoint:
- WebSocket
- HTTP
--  Get
-- ...

File: iroha_http_server/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,640 @@
...
+}
+
+pub type PathParams = BTreeMap<String, String>;
```
Comment (humb1t) [inline iroha_http_server/src/lib.rs:179]: Please read updates to contributing md in my fork and let's try to order our code in some predefined manner

File: iroha_http_server/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,640 @@
...
+    async fn call(
+        &self,
+        state: State,
+        params: PathParams,
+        request: HttpRequest,
+    ) -> Result<HttpResponse, String> {
+        let future = (self)(state, params, request);
+        future.await
+    }
+}
```
Comment (humb1t) [inline iroha_http_server/src/lib.rs:207]: can be moved to http_server::http module

File: iroha_http_server/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,640 @@
...
+    async fn call(
+        &self,
+        state: State,
+        params: PathParams,
+        stream: WebSocketStream,
+    ) -> Result<(), String> {
+        let future = (self)(state, params, stream);
+        future.await
+    }
+}
```
Comment (humb1t) [inline iroha_http_server/src/lib.rs:235]: can be moved to http_server::ws module

File: iroha_http_server/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,640 @@
...
+
+//TODO: Add more response builders.
+#[derive(Debug, Clone, PartialEq)]
+pub struct HttpResponse {
+    /// The response version, such as `HTTP/1.1`.
+    pub version: HttpVersion,
+    /// The response code, such as `200`.
+    pub code: u16,
+    /// The response reason-phrase, such as `OK`.
+    pub reason: String,
```
Comment (humb1t) [inline iroha_http_server/src/lib.rs:357]: May be replaced with enums?

File: iroha_http_server/src/lib.rs (hunk 4 above)
Comment (eadventurous) [reply iroha_http_server/src/lib.rs:179]: Yes, I was working on it at the moment.

File: iroha_http_server/src/lib.rs (hunk 1 above)
Comment (eadventurous) [reply iroha_http_server/src/lib.rs:49]: Used this naming from `tide`. Not a problem to change it, but it just creates a nice flow of `server.at(url).get(handler)`

File: iroha_http_server/src/lib.rs (hunk 2 above)
Comment (eadventurous) [reply iroha_http_server/src/lib.rs:70]: Ok, will check

File: iroha_http_server/src/lib.rs (hunk 7 above)
Comment (eadventurous) [reply iroha_http_server/src/lib.rs]: The reason phrase is not fixed as specified in 6.1.1 section of the [standard](https://www.w3.org/Protocols/rfc2616/rfc2616-sec6.html). So the server might want to use a custom phrase, therefore I think it is better to leave it as `String`.

As for the code itself I think we don't want to waste time specifying all the possible codes at the moment, but the library user might want to use some not very common response, so I also think we should leave it as it is. At least for now.

**Review by humb1t — APPROVED** [review]

**Pull request merged**